### PR TITLE
Count store support for cartesian products

### PR DESCRIPTION
**Pull request by pontusmelke**
For queries such as `MATCH (m),(n),(o),... RETURN count(*)` there is no
need to do the full cartesian product, we can use the count store
directly instead.

### CODE REVIEW

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 1]
```diff
@@ -43,6 +43,12 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
     result.columnAs[Node]("n").toList should equal(List(n1, n2, n3, n4, n5))
   }
 
+  test("foo") {
```
Comment (craigtaverner) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Test name?

Comment (craigtaverner) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Perhaps this was temporary code and should be deleted. This looks superseded by the tests in MatchAggregationsBackedByCountStoreAcceptanceTest.scala.

File: community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/LogicalPlan2PlanDescription.scala [hunk 2]
```diff
@@ -83,7 +83,7 @@ case class LogicalPlan2PlanDescription(idMap: Map[LogicalPlan, Id], readOnly: Bo
         PlanDescriptionImpl(id, "LoadCSV", NoChildren, Seq.empty, variables)
 
       case NodeCountFromCountStore(IdName(variable), labelName, _) =>
-        val arguments = Seq(CountNodesExpression(variable, labelName.map(_.name)))
+        val arguments = Seq(CountNodesExpression(variable, labelName.map(l => l.map(_.name))))
```
Comment (craigtaverner) [inline community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/LogicalPlan2PlanDescription.scala]: Would be better to pluralize the variable name

File: community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planDescription/InternalPlanDescription.scala [hunk 3]
```diff
@@ -116,7 +116,7 @@ object InternalPlanDescription {
     }
     case class ExpandExpression(from: String, relName: String, relTypes:Seq[String], to: String,
                                 direction: SemanticDirection, minLength: Int, maxLength: Option[Int]) extends Argument
-    case class CountNodesExpression(ident: String, label: Option[String]) extends Argument
+    case class CountNodesExpression(ident: String, label: List[Option[String]]) extends Argument
```
Comment (craigtaverner) [inline community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planDescription/InternalPlanDescription.scala]: Would be better to pluralize the parameter name.

**Review by craigtaverner — CHANGES_REQUESTED** [review]

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 1 above)
Comment (pontusmelke) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: yes you are right, I meant to delete this one...

File: community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/NodeCountFromCountStore.scala [hunk 4]
```diff
@@ -23,7 +23,7 @@ import org.neo4j.cypher.internal.frontend.v3_3.ast.LabelName
 import org.neo4j.cypher.internal.ir.v3_3.{CardinalityEstimation, IdName, PlannerQuery}
 
 
-case class NodeCountFromCountStore(idName: IdName, labelName: Option[LabelName], argumentIds: Set[IdName])
+case class NodeCountFromCountStore(idName: IdName, labelName: List[Option[LabelName]], argumentIds: Set[IdName])
```
Comment (craigtaverner) [inline community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/NodeCountFromCountStore.scala:26]: Perhaps pluralize labelNames

**Review by craigtaverner — APPROVED** [review]

**Pull request merged**